File: api/src/routes/scholarships.ts
import { Router } from "express";
import prisma from "../lib/prisma.js";

const router = Router();

router.get("/", async (req, res) => {
  const { studentId } = req.query;
  
  const scholarships = await prisma.scholarship.findMany();
  
  if (studentId) {
    const student = await prisma.student.findUnique({
      where: { id: parseInt(studentId as string) },
      include: { academics: true }
    });
    
    if (student) {
      // Simple rules engine
      // Calculate GPA
      const gradePoints: Record<string, number> = {
        "A": 4.0, "A-": 3.7, "B+": 3.3, "B": 3.0, "B-": 2.7,
        "C+": 2.3, "C": 2.0, "C-": 1.7, "D+": 1.3, "D": 1.0, "F": 0.0
      };
      
      let totalPoints = 0;
      let totalCredits = 0;
      student.academics.forEach(r => {
        const base = gradePoints[r.grade || ""] || 0;
        const weight = r.isAP ? 1.0 : (r.isHonors ? 0.5 : 0.0);
        const credits = r.credits || 1;
        totalPoints += (base + weight) * credits;
        totalCredits += credits;
      });
      const gpa = totalCredits > 0 ? totalPoints / totalCredits : 0;
      
      const matched = scholarships.filter(s => {
        if (s.minGpa && gpa < s.minGpa) return false;
        // Rules engine: "If GPA > 3.5 -> Show 'Academic Excellence Scholarship'"
        if (s.name === "Academic Excellence Scholarship" && gpa <= 3.5) return false;
        return true;
      });
      
      return res.json(matched);
    }
  }
  
  res.json(scholarships);
});

export default router;

File: api_py/routers/scholarships.py
from fastapi import APIRouter, Depends, HTTPException, Request
from ..database import execute_query
from ..auth import get_current_user
from ..utils import calculate_weighted_gpa
from typing import Optional

router = APIRouter()

@router.get("/")
async def get_scholarships(studentId: Optional[str] = None, current_user: dict = Depends(get_current_user)):
    scholarships = execute_query("SELECT * FROM Scholarship")
    
    if studentId:
        student_id = int(studentId)
        # Fetch academic records to calculate GPA for matching
        records = execute_query("SELECT grade, credits, isAP, isHonors FROM AcademicRecord WHERE studentId = %s", (student_id,))
        
        gpa = calculate_weighted_gpa(records)
        
        matched = []
        for s in scholarships:
            if s["minGpa"] and gpa < s["minGpa"]:
                continue
            # Logic: Academic Excellence Scholarship requires GPA > 3.5
            if s["name"] == "Academic Excellence Scholarship" and gpa <= 3.5:
                continue
            matched.append(s)
        return matched
        
    return scholarships
